Portmapper improvements, and options to disable NAT

robmry opened the pull request:
**- What I did**

- Allocate the same port for IPv4 and IPv6.
  - For example:
    - use the same ephemeral port for `0.0.0.0` and `::` in `-p 80`
    - use the same port for all addresses with matching port-mappings (same ranges and protocols):
      - `-p 0.0.0.0:8080-8083:80 -p '[::]:8080-8083:80'`
      - `-p 0.0.0.0::80 -p '[::]::80'`
      - `-p 233.252.0.2::80 -p '[2001:DB8::2]::80`
  - If the same port can't be allocated for all addresses, fail with an error.
  - If a host port is unexpectedly in-use, continue to look for an available port in the rest of the range.
  - fixes https://github.com/moby/moby/issues/42442
  - fixes https://github.com/moby/moby/issues/44136

- Add bridge driver option `com.docker.network.bridge.gateway_mode_ipv6=<nat|routed>` (and `com.docker.network.bridge.gateway_mode_ipv4=<nat|routed>`).
  - When set to `routed`, no NAT or masquerade iptables rules are set up.
  - The container is still firewalled, with an open port for each `-p`.
  - In `inspect` output, a non-NAT'd port mapping is shown with `"HostPort": ""`.
  - So, if the host's network has routing for a docker network, it's possible to communicate directly with the container's address.
  - _As with existing `-p` (`HostConfig.PortBindings`) options, a `--expose` (`Config.ExposedPorts`) for the container's port is still required to make the `-p` do anything. On its own, `--expose` doesn't result in the port being opened._
  - _Host address/ports in `-p` for a disabled-NAT network are ignored, the container's port is opened._

**- How I did it**

- Replaced the v4/v6 portmappers with (simpler) bridge code.
- Also moved the bridge-specific per-port iptables code into the bridge driver.

**- How to verify it**

New/updated tests.

```
# docker network create --ipv6 --subnet fdee::/64 -o com.docker.network.bridge.gateway_mode_ipv6=routed nnn
# netcat -l -p 8080   # make the first port in the range below unusable
# docker run --rm -ti --network nnn --name c1 -p 8080-8083:80 nginx
# docker inspect c1
[
    {
        ...
        "HostConfig": {
            ...
            "PortBindings": {
                "80/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "8080-8083"
                    }
                ]
            },
        "Config": {
            ...
            "ExposedPorts": {
                "80/tcp": {}
            },
        },
        "NetworkSettings": {
            ...
            "Ports": {
                "80/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "8081"
                    },
                    {
                        "HostIp": "::",
                        "HostPort": ""
                    }
                ]
            }
}}]

# IPv4 has NAT rules ...
*nat
-A POSTROUTING -s 172.19.0.2/32 -d 172.19.0.2/32 -p tcp -m tcp --dport 80 -j MASQUERADE
-A POSTROUTING -s 172.19.0.0/16 ! -o br-c68a9a829188 -j MASQUERADE
-A DOCKER ! -i br-c68a9a829188 -p tcp -m tcp --dport 8081 -j DNAT --to-destination 172.19.0.2:80
*filter
-A DOCKER -d 172.19.0.2/32 ! -i br-c68a9a829188 -o br-c68a9a829188 -p tcp -m tcp --dport 80 -j ACCEPT

# IPv6 has no NAT rules ...
*filter
-A DOCKER -d fdee::2/128 ! -i br-c68a9a829188 -o br-c68a9a829188 -p tcp -m tcp --dport 80 -j ACCEPT

# docker-proxy for IPv4 only ...
# ps -ef | grep docker-proxy
root     11894 11420  0 15:08 pts/0    00:00:00 /usr/local/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 8081 -container-ip 172.19.0.2 -container-port 80
```

On a "CentOS 9 (stream)" VM, running firewalld, with addresses `192.168.175.43` and `fd48:70ba:de62:0:a00:27ff:fece:c637` ...
- created a network with no-nat for IPv6 as above (with IPv6 subnet `fd12:3456::/64`).
- Ran `firewall-cmd --reload`, checked that the rules were re-created.
- From another host, added a static route to the VM and checked that  `'http://[fd12:3456::2]'` and `curl 'http://192.168.175.43:8080'` worked as expected.

**- Description for the changelog**
```markdown changelog
- In port mappings with no host port specified, and with host port ranges, allocate the same port for IPv4 and IPv6.
- Add bridge driver option `com.docker.network.bridge.gateway_mode_ipv6=<nat|routed>`, when
  set to `routed` no NAT or masquerading rules are configured for port mappings.
  - This enables direct IPv6 access to the container, if the host's network can route packets for the
     container's address to the host. Mapped ports will be opened in the containers firewall.
  - For example, `docker network create --ipv6 -o com.docker.network.bridge.gateway_mode_ipv6=routed mynet`.
  - Option `com.docker.network.bridge.gateway_mode_ipv4=<nat|routed>` is also available.
```

robmry committed:
Add portallocator.RequestPortsInRange()

Similar to portallocator.RequestPortInRange(), but it attempts to
allocate the same port for multiple IP addresses.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Retry port mapping for a range if ports are in-use

For a port mapping like '-p 8080-8083:80', when some non-docker process
is using a port in the range, try other ports in the range. And, don't
do that on live-restore.

Because the port mapping may fail on live-restore, leaving no ports
mapped for the endpoint - update the view of mapped ports shown in
'inspect' output. (The wrong mappings will still be shown in 'docker ps',
the container will be left running and connected to the network, it just
won't work. There's plenty of scope for better error handling here.)

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Allocate same port for IPv4/IPv6 for 'any interface' mappings.

The bridge driver now does its own port-mapping, rather than using the
portmapper module (which ran as two completely separate instances, for
IPv4 and IPv6).

When asked for a mapping from any host address (0.0.0.0/0) with a range
of host ports, the same port will be allocated for IPv4 and IPv6, or the
mapping will fail with an error if that's not possible.

The bridge driver now manages its own port mappings. So, remove
linux-specific PortMapper code and make what's left Windows-only.

Also, replace the portmapper.userlandProxy interface with StartProxy().

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Option to disable NAT for IPv4/IPv6 for a bridge network.

Add bridge driver options...
  com.docker.network.bridge.gateway_mode_ipv4=<nat|routed>
  com.docker.network.bridge.gateway_mode_ipv6=<nat|routed>

If set to "routed", no NAT or masquerade rules are set up for port
mappings.

When NAT is disabled, the mapping is shown in 'inspect' output with
no host port number. For example, for "-p 80" with NAT disabled for
IPv6 but not IPv4:

    "80/tcp": [
        {
            "HostIp": "0.0.0.0",
            "HostPort": "32768"
        },
        {
            "HostIp": "::",
            "HostPort": ""
        }

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Set up bridge-specific iptables rules in the bridge driver

Use the bridge driver's iptables types to set up portmapping related
iptables rules - instead of using iptables.Forward, which is bridge
specific code in the iptables package.

Remove iptables.Forward() and its unit test, the bridge driver's
version is covered by TestAddPortMappings.

Remove hairpinMode from iptables.ChainInfo hairpinMode relates to bridge
driver specific behaviour, that is now implemented in the bridge driver.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,318 @@
-		return ErrUnsupportedAddressType(fmt.Sprintf("%T", netAddr))
+	res := make([]*portBinding, 0, len(cfg))
+	for _, c := range cfg {
+		pb := &portBinding{PortBinding: c}
```
This should copy 'c', so that if startProxy fails the caller will retry with unmodified config.

robmry committed:
Bind the same port for multiple addresses

Without this change, if a port mapping did not specify a host address
and the network was IPv6-enabled, the same port would be allocated for
mappings from '0.0.0.0' and '::'. But, if the port mapping was specified
with explicit addresses even, for example:
  -p 0.0.0.0:8080-8083:80 -p '[::]:8083-8080:80'

This change looks for port mappings that only differ in the host IP
address, and makes sure it allocates the same port for all of them. If
it can't, it fails with an error.

Signed-off-by: Rob Murray <rob.murray@docker.com>

akerouanton commented on libnetwork/portmapper/mapper.go:
```diff
@@ -1,3 +1,5 @@
+//go:build windows
```
I know that on Linux only the bridge driver uses this package, but honestly I'd prefer to keep both packages clearly separated to make sure we can easily test and evolve the portmapper. For instance, this would be helpful to add nftables support.

Also, looking at it, there are still some bits related to Linux like `StartProxy`, contradicting what the commit message says.

Generally speaking, it's better to not move code and evolve it at the same time. Otherwise it makes your life and reviewers' life harder (eg. when you have to rebase, etc...).

robmry replied:
> I know that on Linux only the bridge driver uses this package, but honestly I'd prefer to keep both packages clearly separated to make sure we can easily test and evolve the portmapper. For instance, this would be helpful to add nftables support.

On Linux yes, only the bridge driver. But there's also Windows - which is almost completely different, but shared the same code. `bridge` had half of the bridge-specific iptables logic, with per-port logic living in `iptables`. We've discussed moving driver-specific logic into the driver several times, it's the right place for it.

The bridge's requirements have changed with the option to have no-NAT, iptables hasn't changed, but the implementation of forwarding/DNAT/masquerading etc was in `iptables.Forward` - behind two layers of indirection in the Linux & Windows `PortMapper` that would also both need to change.

`PortMapper` has a layer of indirection so that Windows can stub out most of the logic. If I'd made the changes in PortMapper, I'd have been changing the Windows code (which is hard to test and doesn't need to change yet, because it doesn't have IPv6 - when it does, the changes needed will inevitably be quite different to what was needed for Linux).

The `PortMapper` was also designed to deal with IPv4/IPv6 separately, which doesn't work.

So, `PortMapper` had to change. The best evolution for it is elimination - with `bridge` composing tools from `portallocator`/`iptables` and `portmapper` docker-proxy code directly.

In a first commit I called `iptables.Forward` directly from `bridge` code instead of indirectly through the `PortMapper`. In a second change I moved its bridge-specific logic into the bridge package - which meant using `bridge.iptRule` to be consistent with the rest of the bridge code. (`bridge.iptRule` itself isn't ideal but it can be improved separately, for all of the `bridge` code that uses it.)

> Also, looking at it, there are still some bits related to Linux like `StartProxy`, contradicting what the commit message says.

Right, the comment says "So, remove linux-specific portmapper code". It should be "So, remove linux-specific PortMapper code".

> Generally speaking, it's better to not move code and evolve it at the same time. Otherwise it makes your life and reviewers' life harder (eg. when you have to rebase, etc...).

Yes, agreed. But I think the alternative to what I've described above is not changing it, or making the messy Windows & Linux implementation more messy, so that it can still be used by both drivers to do quite different things. Is there anything specific I can do to improve things?

akerouanton commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -275,24 +276,96 @@ func (n *bridgeNetwork) setPerPortIptables(b *portBinding, enable bool) error {
+	return nil
+}
+
+func setPerPortNAT(b *portBinding, ipv iptables.IPVersion, proxyPath string, bridgeName string, enable bool) error {
```
The issue with the way you've divided up your commits is that it's not clear what's new code and what's moved from one package to another. This makes review harder, but more importantly, future selves will have a hard time doing archeology.

robmry replied:
Not much has moved. Do you have any concrete suggestions?

robmry replied:
Forgot to note here - but... combined the delete-commits with the others. (I kept them separate to start with, in case there were any concerns about maintaining the old exported package interfaces.)

akerouanton commented on libnetwork/iptables/iptables_test.go:
```diff
@@ -46,59 +44,6 @@ func TestNewChain(t *testing.T) {
 	createNewChain(t)
 }
 
-func TestForward(t *testing.T) {
```
Could you preserve this test while moving code to the `bridge` pkg? Or maybe don't move that code in the first place (see my other comment about that).

robmry replied:
Splitting apart the DNAT/masquerade and filter rules is a main objective of this change, they're now controlled separately... so, the `Forward` function tested here no longer exists. There's better coverage in `TestAllocatePorts`.

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,51 +116,98 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
-	if ip == nil {
```
It's never called with `ip == nil`, and removing the check makes the code simpler. But, I didn't do the same for `RequestPortInRange`, so I can put it back.

robmry replied:
Done.

akerouanton replied:
> It's never called with ip == nil

Oh, right! Then, maybe remove that in a follow-up commit / PR.

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,51 +116,98 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
+	pMaps := make([]*portMap, len(ips))
+	for i, ip := range ips {
+		ipstr := ip.String()
+		protomap, ok := p.ipMap[ipstr]
+		if !ok {
+			protomap = protoMap{
+				"tcp":  p.newPortMap(),
+				"udp":  p.newPortMap(),
+				"sctp": p.newPortMap(),
+			}
+			p.ipMap[ipstr] = protomap
+		}
+		pMaps[i] = protomap[proto]
```
It's unchanged (apart from whitespace), and I've seen code that shape a lot of times. But, sure.

robmry replied:
Done.

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,51 +116,98 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
+			port = pRanges[0].begin
+		}
+
+		available := func() bool {
+			for _, pMap := range pMaps {
+				if _, allocated := pMap.p[port]; allocated {
+					return false
+				}
+			}
+			return true
+		}
+		if available() {
```
Ok. I've been bitten by `slices` and our not-go-modules, but I guess it's ok here.

robmry replied:
Done.

corhere replied:
You were bitten by [some compiler bugs](https://go.dev/issue/66064) which have since been fixed. The `slices` and `maps` packages are all safe for us to use now.

robmry commented on libnetwork/drivers/bridge/bridge_linux_test.go:
```diff
@@ -93,10 +93,23 @@ func TestEndpointMarshalling(t *testing.T) {
+		pb.HostPortEnd = pb.HostPort
+		epms[i] = &pb
+	}
+	if !compareBindings(e.portMapping, epms) {
```
Thank you - I'll fix that.

robmry replied:
Done.

robmry commented on libnetwork/drivers/bridge/bridge_store.go:
```diff
@@ -311,6 +311,17 @@ func (ep *bridgeEndpoint) UnmarshalJSON(b []byte) error {
+	// on live-restore we don't want to allocate different ports - so, remove the range
+	// from the operational data.
+	for _, pm := range ep.portMapping {
+		pm.HostPortEnd = pm.HostPort
```
This is dealing with old saved-state - it represents running config, the ports can't be a range, but that's how the old code stored it.

The new code will treat it as a range on live-restore - it won't know it's live-restore, it's just been given the struct as config as if it had come from the API.

So the old saved-state has to be fixed up ... where else could it be done?

robmry replied:
I've added a comment to say it can be removed when upgrade from 26.x is no longer supported.

robmry commented on libnetwork/drivers/bridge/bridge_store.go:
```diff
@@ -373,17 +373,30 @@ func (ep *bridgeEndpoint) CopyTo(o datastore.KVObject) error {
+// forwarding/NAT rules, and restarts docker-proxy, as needed.
+//
+// TODO(robmry) - if any previously-mapped host ports are no longer available, all
+// iptables forwarding/NAT rules get removed and there will be no docker-proxy
```
There's masquerading too, I generalised it as "NAT" in the comment.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
-		bIPv4 := c.GetCopy()
-		bIPv6 := c.GetCopy()
-		// Allocate IPv4 Port mappings
-		if ok := n.validatePortBindingIPv4(&bIPv4, containerIPv4, defHostIP); ok {
```
I renamed `validatePortBindingIPv4 ` to `configurePortBindingIPv4`, because that's what it does (and did). I don't think there's any less validation ... but, did you spot something specific?

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
-	// Return early if there is a valid Host IP, but its not a IPv4 address
+// configurePortBindingIPv4 returns a new port binding with the HostIP field populated
+// if a binding is required, else nil.
+func configurePortBindingIPv4(bnd types.PortBinding, containerIPv4, defHostIP net.IP) *types.PortBinding {
```
Yes, and it's a pointer to a newly constructed object that just gets stored with other pointers. I guess you prefer an extra "ok" return to indicate that there's no result?

robmry replied:
Done.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
+	for _, c := range cfg {
+		cc := c.GetCopy()
+		pb := &portBinding{PortBinding: &cc}
+		pb.proxy, err = startProxy(c.Proto.String(), c.HostIP, port, c.IP, int(c.Port), proxyPath)
```
Ah, yes - the interface is always 'up' at this point. And for network-connect or live-restore, the container task will already be running. It's probably worth re-jigging, although I think that may leave scope for bugs that vanish iptables rules - I'll take a look.

robmry replied:
I think it's ok as it is ... anything that happens before the iptables rules are added will hit the `FORWARD DROP` action in the filter rule. When the rules are added, the container's port is opened.

For live-restore / network connect, iptables rules are already in place when the proxy is started.

robmry replied:
Discussed with @akerouanton ... the problem is `docker-proxy` will accept a connection that would have passed it by if the iptables rules were in-place. So, although it won't be able to create the container-side connection until the iptables forwarding rule exists, the connection will be abandoned.

It's a small window, and a trade-off between creating rules for a port that belongs to something else (disrupting traffic on an already in-use host port, and possibly breaking it completely when rules are removed).

Perhaps we can modify `docker-proxy` so that it doesn't actually accept connections until it's told the iptables rules are set up ... needs investigation.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
+	if len(containerIPv4) == 0 {
+		return nil
+	}
 	if len(bnd.HostIP) > 0 && bnd.HostIP.To4() == nil {
```
Yes, as I mentioned the other day, our handling of IPv4-mapped IPv6 isn't good. The behaviour hasn't changed, but I'll take a look.

robmry replied:
Addresses are now unmapped early on - and added tests with IPv4-mapped IPv6 default-host, host, and container addresses.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
 		// This change was added to keep backward compatibility
 		containerIP := containerIPv6
-		if ulPxyEnabled && (containerIPv6 == nil) {
+		if proxyPath != "" && (containerIPv6 == nil) {
```
It means `docker-proxy` proxies between a host IPv6 address and a container IPv4 address. And yes, it works.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
+	}
+
+	for _, b := range bindings {
+		if err := n.setPerPortIptables(b, true); err != nil {
```
Just that the iptables rules don't need to be set up if ports can't be reserved - this will change as iptables rules need to be set up first.

robmry replied:
iptables rules don't need to be set up first, so this doesn't have to change. I don't think it needs to change, but can do?

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
+// into mappings for IPv4 and IPv6, because that's how the mapping is presented
+// in 'inspect'). HostPort and HostPortEnd in each returned portBinding are set
+// to the selected and reserved port.
+func (n *bridgeNetwork) allocatePorts(
```
It's the original name, and it's doing exactly what it did before. It's good that that's more obvious now.

It can be renamed. Maybe `addPortMappings`?

robmry replied:
Renamed.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
-		if host, err = portmapper.MapRange(container, bnd.HostIP, int(bnd.HostPort), int(bnd.HostPortEnd), ulPxyEnabled); err == nil {
-			break
+		var b []*portBinding
+		b, err = attemptBindHostPorts(cfg, cfg[0].Proto.String(), cfg[0].HostPort, cfg[0].HostPortEnd, proxyPath)
```
`cfg`, a slice of `PortBinding` is a step between the `PortBinding` that represents config from the API, and the `PortBinding` that represents running config for the API ... it keeps the same form.

This internal function's comment says "the caller is responsible for ensuring that all entries in cfg map the same proto, container port, and host port range (their host addresses must differ)". As long as that holds, it doesn't matter which `cfg` these values are picked up from.

I can add a bunch of extra parameters to duplicate that info.

robmry replied:
It now picks the params out of `cfg[0]`, then checks that the others all match.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
+	} else if b.HostIP.To4() != nil {
+		v = iptables.IPv4
+	} else {
+		// The binding is between containerV4 and hostV6, handled by docker-proxy
```
As above, I'll take a look at the handling of IPv4-mapped IPv6.

robmry replied:
Updated.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,319 @@
-	if bnd.HostIP.To4() == nil {
-		portmapper = n.portMapperV6
+func (n *bridgeNetwork) reapplyPerPortIptables4() {
+	n.reapplyPerPortIptables(func(b *portBinding) bool { return b.IP.To4() != nil })
```
Passing in a filter function is contrived? Better than a bool `ipv4` or similar I think. But, can do that.

corhere reviewed (commented):
All this talk about IPv6-mapped-IPv4 addresses is making me nervous. There is no semantic difference between listen()ing or bind()ing on [::ffff:10.0.0.1]:8080 vs. 10.0.0.1:8080. Only IPv4 packets will be accepted/transmitted over the wire for such a socket. So why does the portmapper need to fuss so much about `net.IP` slice lengths?

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,41 +117,84 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
+//   - it accepts multiple addresses and attempts to find a port in the given range that's
+//     available on each of those addresses. If there is no port that's available for all
+//     addresses, no ports are reserved and an error is returned.
+//   - nil addresses are not replaced by defaultIP
```
Done.

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,41 +117,84 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
-		p.ipMap[ipstr] = protomap
+	pMaps := make([]*portMap, len(ips))
+	for i, ip := range ips {
+		ipstr := ip.String()
+		if _, ok := p.ipMap[ipstr]; !ok {
```
Yes, it'd be better - but also yes, a bit too much for this PR.

robmry commented on libnetwork/portallocator/portallocator.go:
```diff
@@ -116,41 +117,84 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
+	// Starting after the last port allocated for the first address, search
+	// for a port that's available in all ranges.
+	port := pRanges[0].last
+	for i := 0; i <= pRanges[0].end-pRanges[0].begin; i++ {
```
Done.

robmry commented on libnetwork/portmapper/proxy.go:
```diff
@@ -8,7 +8,7 @@ import (
+type UserlandProxy interface {
 	Start() error
 	Stop() error
 }
```
Done ... but "It is only ever called from the StartProxy function" didn't hold before `StartProxy` was added, and the old portmapper code created iptables rules between instantiating and starting the proxy. So, I combined the commits to add `StartProxy` and to make the bridge code use it.

robmry replied:
Except I forgot Windows still needed the interface - so I've made further updates to its code, mostly deleting stuff (but I needn't have combined the commits, argh!).

robmry commented on libnetwork/drivers/bridge/bridge_linux.go:
```diff
@@ -355,6 +352,15 @@ func (n *bridgeNetwork) getNetworkBridgeName() string {
 }
 
+func (n *bridgeNetwork) userlandProxyPath() string {
+	n.Lock()
+	defer n.Unlock()
+	if n.driver == nil {
+		return ""
+	}
```
I think you're right - but every other `'bridgeNetwork` method makes the same checks. A separate review/rework of all that could be good. But, for this change, it seems best to stick with tradition.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,513 @@
+	epAddrV4, epAddrV6 *net.IPNet,
+	cfg []types.PortBinding,
+	defHostIP net.IP,
+) (_ []*portBinding, retErr error) {
```
Done.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,333 @@
+	if len(defHostIP) == 0 {
+		defHostIP = net.IPv4zero
+	} else if addr4 := defHostIP.To4(); addr4 != nil {
+		// Unmap the address if it's IPv4-mapped IPv6.
```
Because iptables vs. ip6tables.

corhere replied:
ip(6)tables does not care either. An IPv4-mapped IPv6 address is a transitional mechanism **for userspace** to be able to write dual-stack apps using only IPv6 APIs. A socket bound to such an address will transmit and receive IPv4 packets over the wire, be filtered by iptables (not ip6tables), etc. Do like the `net` package does and treat 16-byte IPv4 addresses as semantically identical to 4-byte IPv4 addresses. Use the idiom `isIPv4 := addr.To4() != nil` everywhere instead of `isIPv4 := len(addr) == 4` and there is no need to "unmap" anything, or risk bugs if addresses weren't unmapped in some code path or another. I would go so far as to advise _against_ unmapping `net.IP` values here so that any downstream code which incorrectly checks the slice length to determine address family behaves incorrectly in the common case.

robmry replied:
An `iptables` rule with an IPv4-mapped IPv6 address will just-work. It's an IPv4 address.

An `ip6tables` rule with that address can be created, but it'll be useless because the packet it's aiming for will take the IPv4 route.

So, rules need to go into the right tables. The incorrect downstream behaviour you mention would be a bit invisible; the rule would be created normally, it just wouldn't do anything. I don't think that's something we want to invite.

There was one place where I compared lengths of addresses, knowing they were unmapped. It's gone now.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,513 @@
-	container, err := bnd.ContainerAddr()
-	if err != nil {
-		return err
+// bindHostPorts allocates ports and starts docker-proxy for the given cfg, the
```
Done.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,513 @@
+	for _, b := range pm {
+		err = errors.Join(err, n.releasePortBinding(b))
+	}
+	return err
```
Done - by merging with `releasePortBinding` to make `releasePortBindings`, because `releasePortBinding` was called elsewhere but also with a list of `portBinding`.

robmry commented on libnetwork/iptables/iptables_test.go:
```diff
@@ -181,43 +126,6 @@ func TestOutput(t *testing.T) {
-// Runs 10 concurrent rule additions. This will fail if iptables
-// is actually invoked simultaneously without --wait.
-// Note that if iptables does not support the xtable lock on this
-// system, then allowXlock has no effect -- it will always be off.
```
Ah, right - I hadn't (and still haven't) ... but I've put the tests back, now with their own version of what was in `Forward`.

robmry commented on integration/internal/container/ops.go:
```diff
@@ -71,6 +71,16 @@ func WithExposedPorts(ports ...string) func(*TestContainerConfig) {
+	return func(c *TestContainerConfig) {
+		c.HostConfig.PortBindings = nat.PortMap{}
+		for p, b := range pm {
+			c.HostConfig.PortBindings[p] = append([]nat.PortBinding(nil), b...)
```
Done.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -103,6 +117,75 @@ func (n *bridgeNetwork) addPortMappings(
+	}
+	aIP, _ := netip.AddrFromSlice(a.IP)
+	bIP, _ := netip.AddrFromSlice(b.IP)
+	return aIP.Compare(bIP)
```
Done - the order of the addresses isn't actually important, as long as they're grouped together ... but unmapping might remove some future confusion.

robmry commented:
> All this talk about IPv6-mapped-IPv4 addresses is making me nervous. There is no semantic difference between listen()ing or bind()ing on [::ffff:10.0.0.1]:8080 vs. 10.0.0.1:8080. Only IPv4 packets will be accepted/transmitted over the wire for such a socket. So why does the portmapper need to fuss so much about `net.IP` slice lengths?

It needs to put rules into `iptables` or `ip6tables`, so it has to know what sort of address it's really dealing with.

corhere commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,567 @@ func loopbackUp() error {
+		},
+	}
+
+	for _, tc := range testcases {
```
Without a Go 1.21 `//go:build go1.22` constraint on this file, we get Go 1.21 loop variable semantics until we are a Go module.
```suggestion
	for _, tc := range testcases {
	tc := tc	
```

corhere commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,567 @@ func loopbackUp() error {
+			defer netnsutils.SetupTestOSContext(t)()
+
+			// Mock the startProxy function used by the code under test.
+			startProxy = func(proto string,
```
Instead of relying on the returned port mappings to hold references to the mock proxies, keep track of `StartProxy` calls in the test itself, e.g. by tracking the calls in a map keyed by the tuple of arguments and using [`is.DeepEqual`](https://pkg.go.dev/gotest.tools/v3@v3.5.1/assert/cmp#DeepEqual) to assert that all the expected `StartProxy` calls were made and no unexpected calls were included.
```go
type proxyCall struct { proto, host, container, proxyPath string }
proxies := map[proxyCall]*atomic.Bool // proxy -> is not stopped
startProxy = func(...) (func() error, error) {
	c := proxyCall{proto, fmt.Sprintf("%v:%v", hostIP, hostPort), fmt.Sprintf("%v:%v", containerIP, containerPort), proxyPath}
	if _, ok := proxies[c]; ok { return nil, fmt.Errorf("duplicate proxy: %#v", c) }
	proxies[c] = true
	return func() error {
		if tc.expReleaseErr != "" {
			return fmt.Errorf("can't stop now")
		}
		if !proxies[c] {
			return fmt.Errrof("already stopped")
		}
		proxies[c] = false
		return nil
	}
}

robmry replied:
Nice! Done.

corhere commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -1,221 +1,513 @@
+}
+
+func (n *bridgeNetwork) setPerPortIptables(b portBinding, enable bool) error {
+	if len(b.IP) != len(b.HostIP) { // IPv4 addresses have been unmapped.
```
Don't assume that IPv4 addresses have been unmapped. An IPv4 address is an IPv4 address is an IPv4 address once it hits the wire, whether it is encoded in four bytes or in sixteen.
```suggestion
	if (b.IP.To4() != nil) != (b.HostIP.To4() != nil) {
```

robmry replied:
Oh, cool - just did that, then saw this comment.

robmry commented on libnetwork/portmapper/proxy_linux.go:
```diff
@@ -12,12 +12,37 @@ import (
+	hostIP net.IP, hostPort int,
+	containerIP net.IP, containerPort int,
+	proxyPath string,
+) (proxy any, stop func() error, retErr error) {
```
Please could you suggest a way you'll be happy with?

robmry replied:
Oh, missed the comment below about tracking StartProxy calls ... I think it's about the same thing?

akerouanton commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -285,24 +286,96 @@ func (n *bridgeNetwork) setPerPortIptables(b portBinding, enable bool) error {
+	}
+
+	if b.Proto == types.SCTP {
+		// Linux kernel v4.9 and below enables NETIF_F_SCTP_CRC for veth by
```
This isn't needed anymore as the lowest kernel version supported by v27.0 will be 4.18. As a follow-up, we should drop this.

robmry replied:
I've raised https://github.com/moby/moby/issues/47952 to track this.

akerouanton commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,396 @@ func loopbackUp() error {
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc := tc
```
You need to do that outside of `t.Run`.

robmry replied:
Oops, fixed.

robmry commented on libnetwork/drivers/bridge/bridge_linux_test.go:
```diff
@@ -93,10 +93,23 @@ func TestEndpointMarshalling(t *testing.T) {
+	// a different port cannot be selected on live-restore if the original is
+	// already in-use). So, fix up portMapping in the original before running
+	// the comparison.
+	epms := make([]*portBinding, len(e.portMapping))
```
Fixed.

robmry commented on libnetwork/portmapper/proxy_linux.go:
```diff
@@ -12,12 +12,37 @@ import (
+		return nil, nil, err
+	}
+	return proxy, stop, nil
+}
```
Ah, yes - now it doesn't need to `Start()` the returned proxy, that's much better. Done.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,400 @@ func loopbackUp() error {
+			busyPortIPv4: 8080,
+			expPBs: []types.PortBinding{
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 80, HostIP: net.IPv4zero, HostPort: 8081, HostPortEnd: 8081},
+				// Note that, unlike the previous test, IPv4/IPv6 get different host ports.
```
It's not - but it is the expected (old) behaviour.

The final commit, "Bind the same port for multiple addresses", means the two mappings do get the same host port (and it updates this test to match).

robmry commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,400 @@ func loopbackUp() error {
+			epAddrV4: ctrIP4,
+			cfg: []types.PortBinding{
+				{Proto: types.TCP, HostIP: net.IPv4zero, Port: 80},
+				{Proto: types.TCP, HostIP: net.IPv6zero, Port: 80}, // silently ignored
```
If the address is `0.0.0.0`, there doesn't have to be an implicit IPv6 binding.

If there's no proxying and the container has no IPv6, it should error if ...
- the mapping's host address is IPv6, or
- the mapping has no host address, but the default address is IPv6.

So, yes - I'm quite tempted to add the extra validation. But, the silently-ignore is existing behaviour, and I was a bit afraid of missing something.

Perhaps, for now, a TODO comment as well as the new test that points it out is ok? (I'll do that, and can add the validation if there are strong feelings.)

robmry committed:
Remove unused bridge.driver.portAllocator

It was added so that tests could replace it before it was picked
up and used by a new network's PortMapper, so that tests were isolated
from each other. Now the PortMapper is not used by the bridge driver,
neither is driver's portAllocator.

Instead of replacing the driver.portAllocator in tests, reset the
singleton instance using its ReleaseAll().

Un-export portallocator.NewInstance, now the tests aren't using it.

Signed-off-by: Rob Murray <rob.murray@docker.com>

akerouanton commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -66,9 +68,10 @@ func (n *bridgeNetwork) addPortMappings(
-		toBind := make([]portBindingReq, 0, 2)
-		if bindingIPv4, ok := configurePortBindingIPv4(disableNAT4, c, containerIPv4, defHostIP); ok {
+	sortedCfg := sortAndNormPBs(cfg)
+	toBind := make([]portBindingReq, 0, 4)
```
I fail to see why this is 4 and not 2 anymore. Either `-p 80` is used, and both `configurePortBinding...` will append in a single iteration, or `-p 0.0.0.0::80 -p '[::]::80'` is used, and each iteration will append only once.

robmry replied:
I added a comment explaining that, and declared the slice without a size.

robmry commented on libnetwork/portmapper/mapper.go:
```diff
@@ -1,3 +1,5 @@
+//go:build windows
```
I considered that, but decided it'd be better to keep the git history for whichever file would go.

Also, the Windows portmapper will need a similar revamp anyway when we get IPv6 on Windows sorted out. So, most of it's likely to get deleted then anyway.

robmry committed:
Improve PortBinding.String()

Display a PortBinding in a format that's more like the one
used in the CLI, but includes the container IP if known.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Validate port bindings for gateway_mode=routed

When bridge driver opt com.docker.network.bridge.gatway_mode_ipv[46]
is set to "routed", there is no NAT.

When there's no NAT, there's no meaning to the HostPort field in a
port mapping (all the port mapping does is open the container's port),
and the HostIP field is only used to determine the address family.

So, check port bindings, and raise errors if fields are unexpectedly
set when the mapping only applies to a gateway_mode=routed network.
Zero-addresses are allowed, to say the mapping/open-port should be
IPv4-only or IPv6-only, and host ports are not allowed.

A mapping with no host address, so it applies to IPv4 and IPv6 when
the default binding is 0.0.0.0, may include a host port if either
uses NAT. The port number is ignored for the directly-routed family.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,396 @@ func loopbackUp() error {
+				"failed to stop docker-proxy for port mapping tcp/fdf8:b88e:bb5c:3483::2:22/:::2222: can't stop now",
+		},
+	}
+
```
Fixed.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -20,6 +20,11 @@ type portBinding struct {
 }
 
+type portBindingReq struct {
+	*types.PortBinding
```
Un-pointered.

corhere replied:
Nothing's changed?

robmry replied:
Oh, no - I misread the comment, un-pointered `portBinding`, now `portBindingReq` too.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -104,6 +118,75 @@ func (n *bridgeNetwork) addPortMappings(
+// sortAndNormPBs returns a slice of pointers to copies of cfg's entries, where
+// each entry is normalised to have HostPortEnd=HostPort (rather than 0) if the
+// host port isn't a range - sorted into the ordering defined by cmpPortBinding.
+func sortAndNormPBs(cfg []types.PortBinding) []*types.PortBinding {
```
Done.

robmry commented on libnetwork/portmapper/proxy.go:
```diff
@@ -32,28 +27,31 @@ type dummyProxy struct {
 	default:
 		return nil, fmt.Errorf("Unknown addr type: %s", proto)
 	}
+	p := &dummyProxy{addr: addr, ipVersion: version}
```
As they're existing structs, and we may need to add an extra step in the initialisation - after binding, start accepting connections, to deal with https://github.com/moby/moby/pull/47871#discussion_r1624703022 - I don't want to go further down this route at the moment.

robmry commented on libnetwork/portmapper/proxy_linux.go:
```diff
@@ -12,12 +12,31 @@ import (
 	}
 
-	return &proxyCommand{
+	p := &proxyCommand{
```
As above, for `dummyProxy`.

robmry commented:
I've added two commits ...

[Improve PortBinding.String()](https://github.com/moby/moby/pull/47871/commits/ebfe043e8fd87b6fc61453df06d64dd428c4b598) - display port mappings in a format more like the CLI's, with all fields so that it's useful. This doesn't affect `inspect` or anything in the API.

[Validate port bindings for gateway_mode=routed](https://github.com/moby/moby/pull/47871/commits/66639691e18c4e8558463eec1574b13af6cfe16f) - when `-p` is used with `gateway_mode_ipv[46]=routed`, check that host IP/Port aren't specified. Previously they were ignored, but that could be misleading as, for example, `-p 127.0.0.1:8080:80` opened container port `80` for any traffic, it's not restricted to traffic originating from anywhere specific.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux_test.go:
```diff
@@ -171,3 +178,400 @@ func loopbackUp() error {
+				// Check the DNAT rule.
+				dnatRule := ""
+				if tc.proxyPath != "" {
+					// No docker-proxy, so expect "hairpinMode".
```
Raised as https://github.com/moby/moby/issues/47953.
